diff --git a/src/communication/network_interface_cl.c b/src/communication/network_interface_cl.c
index 3ad721619..17ce390f8 100644
--- a/src/communication/network_interface_cl.c
+++ b/src/communication/network_interface_cl.c
@@ -64,6 +64,7 @@
 #include "replication.h"
 #include "es.h"
 #include "db.h"
+#include "db_multi_threads_connections.h"
 #include "db_query.h"
 #include "dbtype.h"
 #include "compile_context.h"
@@ -92,8 +93,8 @@
 
 #if defined (CS_MODE)
 #define NET_DEFER_END_QUERIES_MAX 5
-static QUERY_ID net_Deferred_end_queries[NET_DEFER_END_QUERIES_MAX];
-static int net_Deferred_end_queries_count = 0;
+static CUB_THREAD_LOCAL QUERY_ID net_Deferred_end_queries[NET_DEFER_END_QUERIES_MAX];
+static CUB_THREAD_LOCAL int net_Deferred_end_queries_count = 0;
 #endif /* CS_MODE */
 
 /*
diff --git a/src/transaction/log_applier.c b/src/transaction/log_applier.c
index 0383c2670..b1a8d0d9f 100644
--- a/src/transaction/log_applier.c
+++ b/src/transaction/log_applier.c
@@ -964,16 +964,26 @@ la_reader_commit_apply_info (void)
   int res;
 
   res = la_update_ha_last_applied_info ();
-  if (res < 0)
+  if (res > 0)
     {
-      if (ER_IS_SERVER_DOWN_ERROR (res))
-	{
-	  return ER_NET_CANT_CONNECT_SERVER;
-	}
-      return res;
+      return db_commit_transaction ();
     }
 
-  return db_commit_transaction ();
+  if (res == 0)
+    {
+      la_Info.fail_counter++;
+      er_log_debug (ARG_FILE_LINE, "log applied but cannot update last committed LSA (%d|%d)",
+		    la_Info.committed_lsa.pageid, la_Info.committed_lsa.offset);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_HA_GENERIC_ERROR, 1, "failed to update _db_ha_apply_info");
+      return NO_ERROR;
+    }
+
+  if (ER_IS_SERVER_DOWN_ERROR (res))
+    {
+      return ER_NET_CANT_CONNECT_SERVER;
+    }
+
+  return res;
 }
 
 static void
@@ -6887,7 +6897,6 @@ la_change_state (void)
 static int
 la_log_commit (bool update_commit_time)
 {
-  int res;
   int error = NO_ERROR;
 
   (void) la_find_required_lsa (&la_Info.required_lsa);
@@ -6900,33 +6909,7 @@ la_log_commit (bool update_commit_time)
       la_Info.log_commit_time = time (0);
     }
 
-  error = la_flush_repl_items (true);
-  if (error != NO_ERROR)
-    {
-      return error;
-    }
-
-  res = la_update_ha_last_applied_info ();
-  if (res > 0)
-    {
-      error = la_commit_transaction ();
-    }
-  else
-    {
-      la_Info.fail_counter++;
-
-      er_log_debug (ARG_FILE_LINE, "log applied but cannot update last committed LSA (%d|%d)",
-		    la_Info.committed_lsa.pageid, la_Info.committed_lsa.offset);
-      if (ER_IS_SERVER_DOWN_ERROR (res))
-	{
-	  error = ER_NET_CANT_CONNECT_SERVER;
-	}
-      else
-	{
-	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_HA_GENERIC_ERROR, 1, "failed to update _db_ha_apply_info");
-	  error = NO_ERROR;
-	}
-    }
+  error = la_reader_commit_apply_info ();
 
   return error;
 }
diff --git a/src/transaction/transaction_cl.c b/src/transaction/transaction_cl.c
index d1378e6e8..146b34cf4 100644
--- a/src/transaction/transaction_cl.c
+++ b/src/transaction/transaction_cl.c
@@ -48,6 +48,7 @@
 #include "log_comm.h"
 #include "log_lsa.hpp"
 #include "db_query.h"
+#include "db_multi_threads_connections.h"
 #include "boot_cl.h"
 #include "virtual_object.h"
 #include "schema_manager.h"
@@ -76,7 +77,7 @@ LOCK tm_Tran_rep_read_lock = NULL_LOCK;	/* used in RR transaction locking to not
  * must be set before each transaction command.
  */
 LC_FETCH_VERSION_TYPE tm_Tran_read_fetch_instance_version = LC_FETCH_MVCC_VERSION;
-int tm_Tran_latest_query_status;
+CUB_THREAD_LOCAL int tm_Tran_latest_query_status;
 
 /* Timeout(milli seconds) for queries.
  *
