diff --git a/src/transaction/log_applier.c b/src/transaction/log_applier.c
index fa48a66ba..33547b534 100644
--- a/src/transaction/log_applier.c
+++ b/src/transaction/log_applier.c
@@ -47,6 +47,7 @@
 #include "environment_variable.h"
 #include "message_catalog.h"
 #include "msgcat_set_log.hpp"
+#include "error_context.hpp"
 #include "log_compress.h"
 #include "log_lsa.hpp"
 #include "object_primitive.h"
@@ -743,6 +744,10 @@ static void *
 la_apply_worker_main (void *arg)
 {
   LA_APPLY_WORKER *worker = (LA_APPLY_WORKER *) arg;
+  cuberr::context *er_context_p = NULL;
+
+  er_context_p = new cuberr::context ();
+  er_context_p->register_thread_local ();
 
   while (true)
     {
@@ -790,6 +795,9 @@ la_apply_worker_main (void *arg)
       pthread_mutex_unlock (&worker->mutex);
     }
 
+  er_context_p->deregister_thread_local ();
+  delete er_context_p;
+
   return NULL;
 }
 
